### ISSUE TITLE
0.9.80 broke hs.eventtap.event:post()?

### ISSUE THREAD

**Issue by von** [Hammerspoon/hammerspoon#2498]
I have some code that basically does the following (the snippet in questions send a `;` keypress to the `YT Music` application which plays/pauses playback). Since upgrading to 0.9.80 it does not seem to work any more. The simplified example that follows does not either - no error thrown, it just doesn't seem to do anything. I've tried sending other keystrokes and to other applications with no apparent success.

```
> yt = hs.application.find("YT Music")
> e = hs.eventtap.event.newKeyEvent(";", true)
> e:post(yt)
hs.eventtap.event: Event type: 10 (0x6000019bc938)
```

**Comment by latenitefilms** [comment]
If you go back to Hammerspoon 0.9.78 does it work again? As far as I can see `hs.eventtap.event` hasn't changed for a while. It could also be a Catalina permissions issue so you could try following the instructions here:

https://github.com/Hammerspoon/hammerspoon/issues/2236#issuecomment-568937796

Curious to hear how you go, as I tried send events to a specific application in the past and it didn't work.

**Comment by von** [comment]
I confirmed it does work with 0.9.78.

Reseting the permissions with `tcutil` as described in https://github.com/Hammerspoon/hammerspoon/issues/2236#issuecomment-568937796 does not help.

**Comment by latenitefilms** [comment]
Awesome - thanks heaps for testing! Good to know it's definitely a regression. I'm always using the latest code in the `master` branch, which explains why it didn't work for me when I tried it relatively recently.

I'll have another look at the recent changes to `hs.event` and see what I can see. @asmagill might have some ideas too.

**Comment by latenitefilms** [comment]
I can confirm that this works in 0.9.78 but not in 0.9.80:

```lua
app = hs.application.find("TextEdit")
e = hs.eventtap.event.newKeyEvent("a", true)
e:post(app)
```

I'm assume the issue is due to changes in `hs.uielement`, probably related to [this line](https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/eventtap/event.m#L482):

```objc
AXUIElementRef app = *((AXUIElementRef*)luaL_checkudata(L, 2, "hs.application")) ;
```

Interestingly, the line above it is commented out four years ago ([fix crash when posting event to an application](https://github.com/Hammerspoon/hammerspoon/commit/4f6112677fa6386ade818d3e47e6904f48ced423)):

```objc
AXUIElementRef app = lua_touserdata(L, 2);
```

But I'm wondering now with the changes to `hs.uielement`, whether changing back will help.

I'll give it a shot, unless @asmagill comes back with any better ideas before hand.

**Comment by asmagill** [comment]
The problem is that the `hs.applciation` userdata is no longer a wrapped AXUIElementRef; instead it's now its own `HSapplication` object.

You have two approaches to fix this:

Adopt the approach `hs.application` and `hs.window` do by adding `#import "HSuicore.h"` at the top of the file, then doing something like this in the function itself:

~~~obj-c
HSapplication *appObj = [skin toNSObjectAtIndex:2] ;
AXUIElementRef app = appObj.elementRef ;
~~~

You may need to also link the build against something within the XCode configuration for the subproject, not sure... check out the `window` and `application` subprojects and see if they link against anything other than `LuaSkin`.

Or take the approach I did with `hs.axuielement` (note that this approach was done so I could build it either with *or* without it being part of core:

~~~obj-c
AXUIElementRef getElementRefPropertyFromClassObject(NSObject *object) {
    AXUIElementRef ref      = NULL ;
    SEL            selector = NSSelectorFromString(@"elementRef") ;

    if ([object respondsToSelector:selector]) {
        NSMethodSignature *signature  = [NSMethodSignature signatureWithObjCTypes:"^{__AXUIElement=}16@0:8"] ;
        NSInvocation      *invocation = [NSInvocation invocationWithMethodSignature:signature] ;
        [invocation setTarget:object] ;
        [invocation setSelector:selector] ;
        [invocation invoke] ;
        [invocation getReturnValue:&ref] ;
        if (ref) CFRetain(ref) ;
    }
    return ref ;
}
~~~

Bulkier, but allows for external stand-alone builds. Probably not an issue with `hs.eventtap`.

**Comment by latenitefilms** [comment]
Ok, so I know what the problem is. Looking at [this code](https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/eventtap/event.m#L477):

```objc
static int eventtap_event_post(lua_State* L) {
    CGEventRef event = *(CGEventRef*)luaL_checkudata(L, 1, EVENT_USERDATA_TAG);

    if (luaL_testudata(L, 2, "hs.application")) {
//         AXUIElementRef app = lua_touserdata(L, 2);
        AXUIElementRef app = *((AXUIElementRef*)luaL_checkudata(L, 2, "hs.application")) ;

        pid_t pid;
        AXUIElementGetPid(app, &pid);
```

Using the above test code...

`app` is `<HSapplication: 0x604000754190>`
`pid` is `-1` (ERROR: Unable to get PSN for PID: -1)

...I'll do some digging to work out why this is.

**Comment by latenitefilms** [comment]
Opps, saw your post after I just posted mine. All makes sense - thanks @asmagill. Will test it out!

**Comment by asmagill** [comment]
(Note that if you do take the second approach, you probably don't need the CFRetain line since you're only using and not storing it.)

**Comment by asmagill** [comment]
And this does bring to mind a couple of other places where we used to (and may still) allow converting things like the canvas, console, and webviews into `hs.window` objects that are probably broken as well... I was never happy with it, but it was a quick way to add some functionality without replicating code.

@cmsj, don't do release 81 until I have a chance to see if those instances are still around and address them if they are. They should be seldom used code, but they may still be lurking in there...

**Comment by latenitefilms** [comment]
@asmagill - I've pushed a fix in #2502 which seems to work fine.
